c/misra/src/rules/RULE-5-9/IdentifiersWithInternalLinkageNotUnique.ql

```diff
@@ -0,0 +1,32 @@
+/**
+ * @id c/misra/identifiers-with-internal-linkage-not-unique
+ * @name RULE-5-9: Identifiers that define objects or functions with internal linkage should be unique
+ * @description Using non-unique identifiers can lead to developer confusion.
+ * @kind problem
+ * @precision very-high
+ * @problem.severity error
+ * @tags external/misra/id/rule-5-9
+ *       maintainability
+ *       readability
+ *       external/misra/obligation/advisory
+ */
+
+import cpp
+import codingstandards.c.misra
+
+from Declaration d1, Declaration d2
+where
+  not isExcluded(d1, Declarations6Package::identifiersWithInternalLinkageNotUniqueQuery()) and
+  not isExcluded(d2, Declarations6Package::identifiersWithInternalLinkageNotUniqueQuery()) and
+  d1.isStatic() and
+  d1.isTopLevel() and
+  not d1 = d2 and
+  d1.getName() = d2.getName() and
+  // Apply an ordering based on location to enforce that (d1, d2) = (d2, d1) and we only report (d1, d2).
+  (
+    d1.getFile().getAbsolutePath() < d2.getFile().getAbsolutePath()
+    or
+    d1.getFile().getAbsolutePath() = d2.getFile().getAbsolutePath() and
+    d1.getLocation().getStartLine() < d2.getLocation().getStartLine()
```

Comment:
I think you could implement the deduplication with a simple comparison of the Location rather than having to check both path and line numbers. The implementation does work as-is though.

Reply:
I actually copied it from [here](https://github.com/github/codeql-coding-standards/blob/main/cpp/autosar/src/rules/M8-4-2/NonIdenticalIdentifierUsedForTheParameterInReDeclarationOfAFunction.ql#L35), now that Ive used it a few times though it could be extracted to a predicate... I just was thinking that the general form to accommodate the different times its been used would be with `Locatable`s, in which case that sounds terrible to be evaluated without the context of the query... also though I've used this in like 3 queries now so is that enough to amortize?  lmk what you think!